### PR TITLE
Updated the Pubsub sample

### PR DESCRIPTION
**Pull request by michaelawyu**
As requested in b/36888286.

### CODE REVIEW

**Comment by michaelawyu** [comment]
Hi Frank! Thanks for the review. I have made the requested changes. There is one thing I am not sure though; some comments in the new sample only apply to Ruby (for example, exceptions thrown in callback functions will not propagate to main thread in Ruby Pub/Sub client library), and I feel that it might be OK to include them in the sample? If they should be removed as well, please let me know. Thanks :) 

Chen

File: pubsub/subscriptions.rb [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require "google/cloud/pubsub"
+
+def update_push_configuration project_id:, subscription_name:, new_endpoint:
+  # [START update_push_configuration]
+  # project_id        = Your Google Cloud Project ID
```
Comment (frankyn) [inline pubsub/subscriptions.rb]: Add double quotes around value. 
For example:
`project_id = "Your Google Cloud Project ID"`

File: pubsub/subscriptions.rb [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+end
+
+def delete_subscription project_id:, subscription_name:
+  # [START delete_subscription]
+  # project_id        = Your Google Cloud Project ID
+  # subscription_name = Your Pubsub subscription name
+  pubsub = Google::Cloud::Pubsub.new project: project_id
+
+  subscription = pubsub.subscription subscription_name
+  subscription.delete
```
Comment (frankyn) [inline pubsub/subscriptions.rb]: Add a new line after this line. Try to group sections of the sample based on steps.

File: pubsub/subscriptions.rb [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+end
+
+def get_subscription_policy project_id:, subscription_name:
+  # [START get_subscription_policy]
+  # project_id        = Your Google Cloud Project ID
+  # subscription_name = Your Pubsub subscription name
+  pubsub = Google::Cloud::Pubsub.new project: project_id
+
+  subscription = pubsub.subscription subscription_name
+  policy = subscription.policy
```
Comment (frankyn) [inline pubsub/subscriptions.rb]: Add a new line after this line.

File: pubsub/subscriptions.rb [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require "google/cloud/pubsub"
+
+def update_push_configuration project_id:, subscription_name:, new_endpoint:
+  # [START update_push_configuration]
+  # project_id        = Your Google Cloud Project ID
+  # subscription_name = Your Pubsub subscription name
+  # new_endpoint      = Endpoint where your app receives messages
```
Comment (frankyn) [inline pubsub/subscriptions.rb]: Add require "google/cloud/pubsub" after the commented variables for all samples.

File: pubsub/topics.rb [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+def test_topic_permissions project_id:, topic_name:
+  # [START test_topic_permissions]
+  # project_id = Your Google Cloud Project ID
+  # topic_name = Your Pubsub topic name
+  pubsub = Google::Cloud::Pubsub.new project: project_id
+
+  topic = pubsub.topic topic_name
+  permissions = topic.test_permissions "pubsub.topics.attachSubscription",
+    "pubsub.topics.publish", "pubsub.topics.update"
+  puts permissions.include? "pubsub.topics.attachSubscription"
```
Comment (frankyn) [inline pubsub/topics.rb]: Update lines like this to output that the permission is available.

```ruby
puts "Permission to consume" if permissions.include? "pubsub.subscriptions.consume"
```

File: pubsub/topics.rb [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+end
+
+def create_pull_subscription project_id:, topic_name:, subscription_name:
+  # [START create_pull_subscription]
+  # project_id        = Your Google Cloud Project ID
+  # topic_name        = Your Pubsub topic name
+  # subscription_name = Your Pubsub subscription name
+  pubsub = Google::Cloud::Pubsub.new project: project_id
+
+  topic = pubsub.topic topic_name
```
Comment (frankyn) [inline pubsub/topics.rb]: Align the assignments.

File: pubsub/topics.rb [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+end
+
+def publish_messages_async_with_batch_settings project_id:, topic_name:
+  # [START publish_messages_async_with_batch_settings]
+  # project_id = Your Google Cloud Project ID
+  # topic_name = Your Pubsub topic name
+  pubsub = Google::Cloud::Pubsub.new project: project_id
+
+  topic = pubsub.topic topic_name, async: {
+    # Start sending messages in one request once the size of all queued messages
```
Comment (frankyn) [inline pubsub/topics.rb]: Could you move the comments above the assignment?

```ruby
# Start sending messages...
topic = pubsub.topic topic_name, async: {
```

File: pubsub/subscriptions.rb [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+  pubsub = Google::Cloud::Pubsub.new project: project_id
+
+  subscription = pubsub.subscription subscription_name
+  subscription.endpoint= new_endpoint
```
Comment (frankyn) [inline pubsub/subscriptions.rb]: Add a space to the assignment.

**Comment by michaelawyu** [comment]
Hi Frank! Thanks for the comments. I've made the requested changes :)

File: pubsub/README.md [hunk 9]
```diff
@@ -1,19 +1,121 @@
...
+    YOUR-SUBSCRIPTION
+
+
+## Test samples
+
+Test the sample:
+
+    $ bundle exec rspec
 
 Deploy the push listener:
```
Comment (frankyn) [inline pubsub/README.md]: `Deploy the push listener` should be under Run samples

File: pubsub/subscriptions.rb [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+  # [END test_subscription_permissions]
+end
+
+def listen_for_messages project_id:, subscription_name:
+  # [START listen_for_messages]
+  # project_id        = "Your Google Cloud Project ID"
+  # subscription_name = "Your Pubsub subscription name"
+  require "google/cloud/pubsub"
+
+  pubsub       = Google::Cloud::Pubsub.new project: project_id
```
Comment (frankyn) [inline pubsub/subscriptions.rb]: This line isn't part of a group of assignments could you remove the added space between the variable and the assignment operator?
`pubsub = Google::Cloud::Pubsub.new project: project_id`

File: pubsub/subscriptions.rb [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+
+  subscriber.start
+  # Let the main thread sleep for 60 seconds so the thread for listening
+  # messages does not quit
+  sleep 60
+  subscriber.stop.wait!
+  # [END listen_for_messages_with_concurrency_control]
+end
+
+if __FILE__ == $0
```
Comment (frankyn) [inline pubsub/subscriptions.rb]: Substitute $0 for $PROGRAM_NAME.

File: pubsub/topics.rb [hunk 12]
```diff
@@ -0,0 +1,339 @@
...
+
+def list_topic_subscriptions project_id:, topic_name:
+  # [START list_topic_subscriptions]
+  # project_id = "Your Google Cloud Project ID"
+  # topic_name = "Your Pubsub topic name"
+  require "google/cloud/pubsub"
+
+  pubsub = Google::Cloud::Pubsub.new project: project_id
+
+  topic  = pubsub.topic topic_name
```
Comment (frankyn) [inline pubsub/topics.rb]: Align assignments here.

File: pubsub/topics.rb [hunk 13]
```diff
@@ -0,0 +1,339 @@
...
+  # [START create_push_subscription]
+  # project_id        = "Your Google Cloud Project ID"
+  # topic_name        = "Your Pubsub topic name"
+  # subscription_name = "Your Pubsub subscription name"
+  # endpoint          = "Endpoint where your app receives messages"
+  require "google/cloud/pubsub"
+
+  pubsub = Google::Cloud::Pubsub.new project: project_id
+
+  topic  = pubsub.topic topic_name
```
Comment (frankyn) [inline pubsub/topics.rb]: Align assignments here.

**Comment by michaelawyu** [comment]
Hi Frank! Thanks for helping me out :) I've made the requested changes. Topics.rb line #55 and #156 has been updated as well, not sure why GitHub won't recognize the old snapshots as outdated :(

**Comment by frankyn** [comment]
Thanks @michaelawyu LGTM! Is documentation updated for these samples? cloud.google.com/pubsub/docs?

**Comment by michaelawyu** [comment]
Thanks :) I've created a CL. Will publish the page to the staging site for tech writers to review once the new sample is published.

**Comment by frankyn** [comment]
@michaelawyu sg! I wanted to make sure, so we will wait to merge until the CL is reviewed and ready to submitted.

**Comment by michaelawyu** [comment]
Hi Frank! The CL (cl/186061185) has been approved. If it is OK with you, I will merge the sample and submit the CL after the final check. Thanks!

**Comment by frankyn** [comment]
Hi @michaelawyu! I left a comment in the CL before we merge this I'd like for you to stage the documentation before publishing/merging cl and PR.

**Pull request merged**